kernel/byterun/coq_interp.c

```diff
@@ -957,6 +959,7 @@ value coq_interprete
         print_int(sizes & 0xFFFFFF);
         if (Is_block(accu)) {
           long index = Tag_val(accu);
+          if (index == Closure_tag) index = 0;
```

Comment:
@bgregoir says: how about writing `index = (index == Closure_tag) ? 0 : index`

Reply:
How is that an improvement? I find it less readable.

Or perhaps you fear that the compiler cannot optimize the code otherwise? Then, fear not, as the produced assembly is the following branchless conditional on x86:

```asm
cmpq	$247, %rax
cmovne	%rax, %rcx
```

In fact, it is almost unfortunate that this conditional is optimized, because a branching conditional would be correctly predicted here. Eventually, we will have to perform a profile-guided compilation of the bytecode interpreter.

Reply:
I think @bgregoir suggested it mainly so that it is optimized.

Reply:
@silene, you are write, I want to have a CMOVXX instruction. So since the compiler use it, I'm fine with your version. 
For your remark with conditional branching and branch prediction our experiments to obtain "constant-time" code 
show also that it is generally more efficient to use CMOV instead of branching. I think it is also why the compiler use it.

Reply:
The thing is, we do not want it to be constant time, we want it to be zero time. (The fastest instruction is the one that is not executed.) As Intel's documentation states it, "Use the `SETCC` and `CMOV` instructions to eliminate unpredictable conditional branches where possible. Do not do this for predictable branches. [...] In addition, converting a conditional branch to `SETCC` or `CMOV` trades off control flow dependence for data dependence and restricts the capability of the out-of-order engine." Here, the branch is predictable when the user computes a closed term, which hopefully is the main use case for `vm_compute`.

Reply:
Another idea would be to have a simple static analysis in the compiler detect closed terms and generate a code calling only the `NOCHECK` operations, statically saving all those tests.

Reply:
I have never say that we want constant-time code ... Anyway, I am sur that using a conditional jump or a CMOV will not change the complexity of the evaluation, and since the compiler does what I want :-), I am happy with keeping this line as you want. For proux01, we have frequently discuss this solution with Maxime, we did not implement it first because you have to duplicate the code, but also because a function can be a closed term and to normalise it you have to normalise its body which is not closed anymore. I agree that this last argument is not very good, because in practice the result is a closed term in a simple datatype (without any function inside).